src/main/scala/com/fulcrumgenomics/bam/api/SamRecord.scala

```diff
@@ -236,9 +236,83 @@ trait SamRecord {
   // transient attributes
   @inline final val transientAttrs: TransientAttrs = new TransientAttrs(this)
 
-  // TODO long-term: replace these two methods with methods on [[Cigar]] to save creating alignment blocks in memory
-  @inline final def refPosAtReadPos(pos: Int) = getReferencePositionAtReadPosition(pos)
-  @inline final def readPosAtRefPos(pos: Int, returnLastBaseIfDeleted: Boolean) = getReadPositionAtReferencePosition(pos, returnLastBaseIfDeleted)
+  /** Returns the 1-based reference position for the 1-based position in the read, or [[None]] if the reference does
+    * not map at the read position (e.g. insertion).
+    *
+    * @param pos the 1-based read position to query
+    * @param returnLastBaseIfInserted if the reference is an insertion, true to returned the previous reference base,
+    *                                 false to return None
+    * */
+  @inline final def refPosAtReadPos(pos: Int, returnLastBaseIfInserted: Boolean = false): Option[Int] = if (pos <= 0) None else {
+    var readPos      = 1
+    var refPos       = this.start
+    var elementIndex = 0
+    val elems        = this.cigar.elems
+
+    def continue(): Boolean = if (elementIndex >= elems.length) false else {
+      val elem = elems(elementIndex)
+      if (pos > CoordMath.getEnd(readPos, elem.lengthOnQuery)) true // current cigar element is before the desired position
+      else {
+        if (pos < readPos) readPos = 0 // past the point in the read
+        else if (elem.operator == CigarOperator.INSERTION) {
+          if (!returnLastBaseIfInserted) refPos = 0
+          else refPos -= 1
+        } // in an insertion, no reference position
+        else refPos += pos - readPos  // get the offset
```

Comment:
I think you need to also handle soft-clipping here to return None.

Reply:
And I think before returning false you need to also check that the cigar operator consumes query bases.  E.g. I think if you have a cigar like `20M10D20M` with `rec.start == 100` and you ask for `rec.refPosAtReadPos(21)` I think this will erroneously return 121 instead of 131.

Reply:
1. I think it handles soft-clipped bases just fine, both shown by the added tests, as well as that the `refPos` will now be past `this.end`!

Reply:
added a comment